File: 02_rookie/week12/234_palindrome-linked-list_watermelon.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    let fast = head;
    let slow = head;
    let reverse = null;

    while (fast && fast.next) {
        fast = fast.next.next;
        prev_reverse = reverse // reverse 저장
        next_slow = slow.next // slow.next 저장
        reverse = slow;
        reverse.next = prev_reverse;
        slow = next_slow;
    }

    if (fast) {
        slow = slow.next;
    }

    while (slow && slow.val === reverse.val) {
        slow = slow.next;
        reverse = reverse.next;
    }
    return !slow;
};

File: 02_rookie/week12/234_palindrome-linked-list_watermelon.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        entire_list = [] # 연결 리스트의 모든 값을 저장할 리스트
        while head:
            entire_list.append(head.val)
            head = head.next

        return entire_list == entire_list[::-1]
